# Car Price Prediction — Data Understanding

## Notebook 01: Dataset Loading and Initial Inspection

**Author:** Muhammad Waqas  
**Project:** Linear and Regularized Regression  
**Dataset Source:** Kaggle  

### Objective

The objective of this notebook is to understand the structure, features, target variable, quality, and initial characteristics of the car price dataset before preprocessing or model training.

## 1. Import Required Libraries

In this section, the libraries required for data loading, numerical operations, file-path management, and visualization are imported.

In [2]:
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## 2. Define and Verify the Dataset Path

Before loading the dataset, its file path is defined and verified to ensure that Python can locate the correct raw data file.

In [4]:
DATA_PATH = Path("../data/raw/car_price_prediction.csv")


print("File exists:", DATA_PATH.exists())

File exists: True


## 3. Load the Dataset

The raw car price dataset is loaded into a Pandas DataFrame for inspection and analysis. No modification is made to the original data at this stage.

In [5]:
df = pd.read_csv(DATA_PATH)

print("Dataset loaded successfully.")

Dataset loaded successfully.


## 4. Preview the Dataset

The first five observations are displayed to understand the dataset columns, values, formatting, and possible target variable.

In [6]:
df.head()

,Car ID,Brand,Year,Engine Size,Fuel Type,Transmission,Mileage,Condition,Price,Model
0,1,Tesla,2016,2.3,Petrol,Manual,114832,New,26613.92,Model X
1,2,BMW,2018,4.4,Electric,Manual,143190,Used,14679.61,5 Series
2,3,Audi,2013,4.5,Electric,Manual,181601,New,44402.61,A4
3,4,Tesla,2011,4.1,Diesel,Automatic,68682,New,86374.33,Model Y
4,5,Ford,2009,2.6,Diesel,Manual,223009,Like New,73577.10,Mustang


## 5. Check Dataset Dimensions

The number of observations and columns is checked to understand the initial dimensions of the dataset.

In [7]:
number_of_rows, number_of_columns = df.shape

print(f"Number of observations (rows): {number_of_rows:,}")
print(f"Number of columns: {number_of_columns}")

Number of observations (rows): 2,500
Number of columns: 10


## 6. Inspect Column Names

The dataset columns are listed to identify the available variables and their possible roles as identifiers, input features, or the prediction target.

In [8]:
print("Dataset columns:\n")

for number, column in enumerate(df.columns, start=1):
    print(f"{number}. {column}")

Dataset columns:

1. Car ID
2. Brand
3. Year
4. Engine Size
5. Fuel Type
6. Transmission
7. Mileage
8. Condition
9. Price
10. Model


## 7. Define the Target and Candidate Features

`Price` is selected as the prediction target. `Car ID` is treated as an identifier because it represents a record number rather than a meaningful car characteristic. The remaining columns are considered candidate input features.

In [9]:
target = "Price"

features = df.columns.drop(["Car ID", "Price"])

print("Target:", target)
print("Features:", list(features))
print("Number of features:", len(features))

Target: Price
Features: ['Brand', 'Year', 'Engine Size', 'Fuel Type', 'Transmission', 'Mileage', 'Condition', 'Model']
Number of features: 8


## 8. Inspect Dataset Structure and Data Types

The dataset structure is inspected to identify the number of records, data types, non-null values, and memory usage.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Car ID        2500 non-null   int64  
 1   Brand         2500 non-null   object 
 2   Year          2500 non-null   int64  
 3   Engine Size   2500 non-null   float64
 4   Fuel Type     2500 non-null   object 
 5   Transmission  2500 non-null   object 
 6   Mileage       2500 non-null   int64  
 7   Condition     2500 non-null   object 
 8   Price         2500 non-null   float64
 9   Model         2500 non-null   object 
dtypes: float64(2), int64(3), object(5)
memory usage: 195.4+ KB


## 9. Check Missing Values

Missing values are checked because incomplete observations can affect statistical analysis, preprocessing, and model performance.

In [11]:
df.isnull().sum()

Car ID          0
Brand           0
Year            0
Engine Size     0
Fuel Type       0
Transmission    0
Mileage         0
Condition       0
Price           0
Model           0
dtype: int64

## 10. Check Duplicate Records

Duplicate records are checked to determine whether the dataset contains repeated observations that could influence the analysis or model training.

In [12]:
df.duplicated().sum()

np.int64(0)

### Initial Data Quality Observation

The dataset contains no missing values and no exact duplicate records. However, further checks are required for identifier uniqueness, categorical consistency, numerical ranges, and potential outliers before considering the data suitable for model training.

## 11. Check Identifier Uniqueness

The `Car ID` column is checked to confirm whether each observation has a unique identifier.

In [13]:
print("Total rows:", len(df))
print("Unique Car IDs:", df["Car ID"].nunique())

Total rows: 2500
Unique Car IDs: 2500


### Identifier Observation

The `Car ID` column contains 2,500 unique values for 2,500 observations. It is a valid record identifier but does not represent a meaningful car characteristic. Therefore, it will be preserved in the raw dataset and excluded from model training.

## 12. Identify Numerical and Categorical Variables

The columns are grouped according to their data types. This distinction is important because numerical and categorical variables require different preprocessing methods.

In [16]:
numerical_columns = df.select_dtypes(include="number").columns
categorical_columns = df.select_dtypes(include="object").columns

print("Numerical columns:", list(numerical_columns))
print("Categorical columns:", list(categorical_columns))

Numerical columns: ['Car ID', 'Year', 'Engine Size', 'Mileage', 'Price']
Categorical columns: ['Brand', 'Fuel Type', 'Transmission', 'Condition', 'Model']


### Feature Type Observation

The numerical input features are `Year`, `Engine Size`, and `Mileage`. The categorical input features are `Brand`, `Fuel Type`, `Transmission`, `Condition`, and `Model`. `Car ID` is an identifier, while `Price` is the continuous prediction target.

## 13. Examine Numerical Summary Statistics

Descriptive statistics are calculated for the meaningful numerical variables to understand their central values, spread, ranges, and possible unusual observations.

In [20]:
df[["Year", "Engine Size", "Mileage", "Price"]].describe().round(1)

,Year,Engine Size,Mileage,Price
count,2500.0,2500.0,2500.0,2500.0
mean,2011.6,3.5,149749.8,52638.0
std,7.0,1.4,87920.0,27295.8
min,2000.0,1.0,15.0,5011.3
25%,2005.0,2.2,71831.5,28908.5
50%,2012.0,3.4,149085.0,53485.2
75%,2018.0,4.7,225990.5,75838.5
max,2023.0,6.0,299967.0,99982.6


### Numerical Summary Observation

All selected numerical variables contain 2,500 observations, confirming that they have no missing values.

- The vehicle years range from 2000 to 2023, with a median year of 2012.
- Engine size ranges from 1.0 to 6.0, with an average of approximately 3.5.
- Mileage ranges from 15 to 299,967, indicating the presence of both very low- and high-mileage vehicles.
- Car prices range from approximately USD 5,011 to USD 99,983.
- The average car price is approximately USD 52,638, while the median is approximately USD 53,485.
- The price standard deviation of approximately USD 27,296 indicates substantial variation in vehicle prices.

The observed ranges do not immediately show impossible negative values. However, distributions and possible outliers must be examined visually before drawing final conclusions.

## 14. Examine Categorical Summary Statistics

Categorical variables are summarized to understand their number of categories, most frequent values, and frequency distributions.

In [21]:
df[["Brand", "Fuel Type", "Transmission", "Condition", "Model"]].describe()

,Brand,Fuel Type,Transmission,Condition,Model
count,2500,2500,2500,2500,2500
unique,7,4,2,3,28
top,Toyota,Diesel,Manual,Used,Fiesta
freq,374,655,1308,855,103


### Categorical Summary Observation

All categorical variables contain 2,500 observations and therefore have no missing values.

- The dataset contains 7 car brands.
- Fuel type has 4 categories.
- Transmission has 2 categories.
- Vehicle condition has 3 categories.
- The model column contains 28 categories and has the highest categorical cardinality.
- Toyota is the most frequent brand, Diesel is the most frequent fuel type, Manual is the most frequent transmission, Used is the most frequent condition, and Fiesta is the most frequent model.

Category labels and complete frequency distributions must be examined before encoding to detect possible spelling inconsistencies or category imbalance.

## 15. Inspect Unique Category Labels

The unique labels of each categorical variable are inspected to detect inconsistent spelling, unexpected categories, or formatting problems.

In [23]:
print("Brands:", df["Brand"].unique())
print("Fuel Types:", df["Fuel Type"].unique())
print("Transmissions:", df["Transmission"].unique())
print("Conditions:", df["Condition"].unique())
print("Models:", df["Model"].unique())

Brands: ['Tesla' 'BMW' 'Audi' 'Ford' 'Honda' 'Mercedes' 'Toyota']
Fuel Types: ['Petrol' 'Electric' 'Diesel' 'Hybrid']
Transmissions: ['Manual' 'Automatic']
Conditions: ['New' 'Used' 'Like New']
Models: ['Model X' '5 Series' 'A4' 'Model Y' 'Mustang' 'Q7' 'Q5' 'Civic'
 'Explorer' 'Model 3' 'Fiesta' 'X3' 'GLA' 'A3' 'X5' 'C-Class' 'E-Class'
 'CR-V' 'Camry' 'Accord' 'GLC' 'Corolla' 'Fit' 'Model S' 'Prius'
 '3 Series' 'RAV4' 'Focus']


### Category Label Observation

The categorical labels appear consistently formatted and do not show obvious spelling or capitalization problems. Fuel type, transmission, and vehicle condition contain clear and meaningful categories. Brand and model names are also readable and consistently represented.

No category merging or renaming is required at this stage. However, category frequencies and the logical relationship between brands and models still require further examination.

## 16. Initial Research Questions

This project will investigate the following questions:

1. Which vehicle characteristics have the strongest relationship with car price?
2. How do manufacturing year, engine size, and mileage influence price?
3. Does car price differ across brands, models, fuel types, transmissions, and vehicle conditions?
4. Are the relationships between numerical features and car price linear or nonlinear?
5. Are there unusual observations or outliers that may influence regression models?
6. Does regularization improve model stability and test performance compared with ordinary linear regression?
7. Which model provides the best balance between prediction performance and interpretability?

## 17. Initial Dataset Understanding Summary

The car price dataset contains 2,500 observations and 10 columns. `Price` is the continuous prediction target, while `Car ID` is a unique identifier that will be excluded from model training.

The dataset contains three meaningful numerical input features: `Year`, `Engine Size`, and `Mileage`. It also contains five categorical input features: `Brand`, `Fuel Type`, `Transmission`, `Condition`, and `Model`.

No missing values or exact duplicate records were identified. The `Car ID` column contains a unique value for every observation. The categorical variables contain consistently formatted labels, with 7 brands, 4 fuel types, 2 transmission types, 3 vehicle conditions, and 28 car models.

Car prices range from approximately USD 5,011 to USD 99,983, with an average price of approximately USD 52,638. The wide variation in price supports the need for exploratory analysis and regression modeling.

The next notebook will examine feature distributions, category frequencies, relationships with price, correlations, and potential outliers.